### PR TITLE
Allow `loop` in constant evaluation

### PR DESCRIPTION
**Pull request by oli-obk**
[Rendered](https://github.com/oli-obk/rfcs/blob/const_looping/text/0000-const-looping.md)

### CODE REVIEW

File: text/0000-const-looping.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+- Start Date: 2018-02-18
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow the use of `loop`, `while` and `while let` during constant evaluation.
+`for` loops are technically allowed, too, but can't be used in practice because
+each iteration calls `iterator.next()`, which is not a `const fn` and thus can't
```
Comment (Centril) [inline text/0000-const-looping.md]: Linking #2237 which (or a modified version of which) will fix that and let you use `iter.next()` in `const fn`.

File: text/0000-const-looping.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+be called within constants.
+
+# Motivation
+[motivation]: #motivation
+
+Any iteration is expressible as a recursion. Since we already allow recursion
```
Comment (Centril) [inline text/0000-const-looping.md]: Small possible improvement: s/a recursion/with recursion.

File: text/0000-const-looping.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+via const fn and termination of said recursion via `if` or `match`, all code
+enabled by const recursion is already legal now. Writing loops with recursion is
```
Comment (Centril) [inline text/0000-const-looping.md]: Does this hold given recursion depth limits?

File: text/0000-const-looping.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+very tedious and can quickly become unreadable, while regular loops are much
+more natural in Rust.
```
Comment (Centril) [inline text/0000-const-looping.md]: This feels like a somewhat **bold** assertion that seems to suggest that functional languages are less readable, while I'd argue that the opposite is the case, recursion is often more readable than loops - and iteration combinators using `map`, `filter`, `fold` even more so (tho combinators have nothing to do with recursion).

I agree that regular loops are more used in Rust, but I believe this has more to do with the lack of guaranteed TCO (Tail Call Optimization, https://en.wikipedia.org/wiki/Tail_call, https://stackoverflow.com/questions/310974/what-is-tail-call-optimization) which RFC #1888 attempts to address.

I don't think any of these assertions are necessary to motivate loops in `const fn` - feature parity alone and the feeling that the language is uniform when you switch from `fn -> const fn` should be enough.

File: text/0000-const-looping.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+If you previously had to write functional code inside constants, you can now
+change it to imperative code. For example if you wrote a fibonacci like
```
Comment (Centril) [inline text/0000-const-looping.md]: Who doesn't love a fibonacci example =P This section feels very nicely written!
Stylistically I'ld do: `s/like/like:` (and before the other sections as well..)

File: text/0000-const-looping.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        i += 1;
+    }
+    b
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A loop in MIR is a cyclic graph of BasicBlocks. Evaluating such a loop is no
```
Comment (Centril) [inline text/0000-const-looping.md]: s/BasicBlocks/`BasicBlocks`

File: text/0000-const-looping.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+different from evaluating a linear sequence of BasicBlocks, except that
+termination is not guaranteed. To ensure that the compiler never hangs
+indefinitely, we count the number of terminators processed and once we reach a
+fixed limit, we report an error mentioning that we aborted constant evaluation,
```
Comment (Centril) [inline text/0000-const-looping.md]: Perhaps a word about if and how you can control that fixed limit? (I assume the `#[recursion_limit = "<limit>"]` attribute?)

File: text/0000-const-looping.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+because we could not guarantee that it'll terminate.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Loops are not guaranteed to terminate
```
Comment (Centril) [inline text/0000-const-looping.md]: Nicely written.

In the **far future** it might be interesting to have a `total` effect in the language with a totality checker that guarantees termination.

File: text/0000-const-looping.md (hunk 3 above)
Comment (oli-obk) [reply text/0000-const-looping.md]: yea, if the loop takes too many iterations you'll hit the limit, but that's not much of an issue imo, since you can always work around that with unrolling or similar tricks.

File: text/0000-const-looping.md (hunk 7 above)
Comment (oli-obk) [reply text/0000-const-looping.md]: there's no such attribute

Comment (Centril) [reply text/0000-const-looping.md]: Not this? https://github.com/rust-lang/rust/blob/master/src/librustc/middle/recursion_limit.rs

Comment (oli-obk) [reply text/0000-const-looping.md]: That's a different recursion. Recursing in const eval does not actually cause recursion in the compiler, just more elements in the virtual stack `Vec`.

File: text/0000-const-looping.md (hunk 3 above)
Comment (Centril) [reply text/0000-const-looping.md]: Ah, thanks for explaining =)

Still... might it be good to allow users to raise the limit from the default so that it is not so fixed and so that you can fix cases wherein you know that termination is guaranteed but rustc complains nonetheless?

File: text/0000-const-looping.md (hunk 8 above)
Comment (joshtriplett) [reply text/0000-const-looping.md]: Do we have a mechanism to adjust the amount of compile-time evaluation allowed, similar to the macro limits?

Comment (oli-obk) [reply text/0000-const-looping.md]: Nope, but it's trivial to add (a single line next to the recursion limit attr code)

**Comment by scottmcm** [comment]
Just counting all terminators worries me slightly that a tweak to MIR optimizations or HIR->MIR lowering will result in code that happened to be right on the edge breaking in an update.

Is there some tweak to the rule that's more resilient to otherwise-invisible changes, and where it's easier to understand by looking at your code how many of whatever it's counting you're taking?  Like maybe only counting backedges taken.  (Though even that changes for loop unrolling, I guess.)

Come to think of it, I wonder if the recursion limit means that the MIR inliner should add a new marker statement to the MIR so that heuristic changes don't change how many times a function recurses...

**Comment by oli-obk** [comment]
Well.... the current limit is riddiculously high. 1 million terminators take a while to evaluate.

We could also hash the entire state at every terminator and detect true loops this way. We could even only do that once we reach the limit so we don't slow down the regular evaluations.

**Comment by oli-obk** [comment]
We could emit a warning for long evaluations so ppl know rustc isn't hanging, just busy

**Comment by burdges** [comment]
Also test builds could report how much time was spent in rustc itself vs const evaluation, so slow crates get blamed quickly.

**Comment by Centril** [comment]
> Well.... the current limit is riddiculously high. 1 million terminators take a while to evaluate.

Do you have some guess as to how much *a while* might be? Just so we're on the same page =)

**Comment by oli-obk** [comment]
While that depends on the code you're executing, experience with when the limit was on statements was several minutes

**Comment by nagisa** [comment]
We should *not* have any error on constant evaluation by default. Not time and not terminator count based. Instead, there should exist a lint of-sorts that would warn every so often that such-and-such constant item is still being evaluated, and that’s it. People who want to make sure stuff doesn’t get out of hand can then #[deny] it and people who know they gonna do heavy compile-time computations can #[allow] it so the compiler doesn’t nag at them too much.

So basically what [this](https://github.com/rust-lang/rfcs/pull/2344#issuecomment-368227609) says.

**Comment by joshtriplett** [comment]
@nagisa I'd be concerned about people running automated builds and finding that they take forever, in that case. CI, for instance.

**Comment by nagisa** [comment]
@joshtriplett Sure, hence `#[deny]`. Of course we still would need to provide a way to configure actual limit. Most importantly, with this approach breaking code which uses `#[deny(long_const_eval)]` would be about as breaking as breaking code which specifies `#[deny(warnings)]` (we do not consider it to be a breaking change).

**Comment by clarcharr** [comment]
Considering how Rust's type system is turing complete, infinite compile loops are already possible. This just makes them a bit easier.

**Comment by oli-obk** [comment]
To the best of my knowledge you can't do infinite loops in the type system without hitting the recursion limit. True recursions would trigger an error of the form "needed X to compute X"

**Comment by oli-obk** [comment]
They are different, because now there can be repetitive recursions. I think going with the warning and no limit is the way to go. I'll change the RFC accordingly

**Comment by oli-obk** [comment]
Long running const eval doesn't cause errors anymore: https://github.com/rust-lang/rust/issues/49980

File: text/0000-const-looping.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+termination is not guaranteed. To ensure that the compiler never hangs
+indefinitely, we count the number of terminators processed and whenever we reach
+a fixed limit, we report a lint mentioning that we cannot guarantee that the
+evaluation will terminate and reset the counter to zero. This lint should recur
+in a non-annoying amount of time (e.g. at least 30 seconds between occurrences).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Infinite loops will hang the compiler if the lint is not denied
```
Comment (pnkfelix) [inline text/0000-const-looping.md]: Nit: Even just a `#[warn]` on the lint will cause warning output to occur (and repeat ad nauseaum) in response to an infinite loop, right? Usually when I hear "hang" I think of "no output visible"... so I would personally have written "if the lint is `#[allow]`'ed" here. But that might just be my own personal interpretation.

File: text/0000-const-looping.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A loop in MIR is a cyclic graph of `BasicBlock`s. Evaluating such a loop is no
+different from evaluating a linear sequence of `BasicBlock`s, except that
+termination is not guaranteed. To ensure that the compiler never hangs
+indefinitely, we count the number of terminators processed and whenever we reach
+a fixed limit, we report a lint mentioning that we cannot guarantee that the
+evaluation will terminate and reset the counter to zero. This lint should recur
+in a non-annoying amount of time (e.g. at least 30 seconds between occurrences).
```
Comment (pnkfelix) [inline text/0000-const-looping.md]: So just to be clear: the const-eval state will include both a terminator-counter and a timestamp; When the terminator-counter both 1. exceeds the (here unspecified) fixed limit, and 2. the current time delta from the timestamp is non-annoying, then we report the lint, reset the terminator-counter *and* timestamp, and finally resume const-eval execution?

(I'm just checking that there's not a typo here and that we are planning on using both pieces of state for the filter here, something of which I am in favor.)

**Comment by pnkfelix** [comment]
@rfcbot fcp merge

(I usually construct a comment thread summary before issuing the rfcbot command, but in this case I *think* all of the relevant comments were addressed, usually via incorporation into the RFC text, so I'm able to just recommend that people read the RFC.)

File: text/0000-const-looping.md (hunk 10 above)
Comment (oli-obk) [reply text/0000-const-looping.md]: The current implementation just has a MIR terminator counter for the warning. There's no host time check. Shouldn't be too hard to add.

**Comment by Centril** [comment]
**Huzzah!** This RFC is now **merged!**

Tracking issue: https://github.com/rust-lang/rust/issues/52000
